V1.0.0.0rc (Project Catalyst) (#8)
fallen-icarus/cardano-loans: README.md

likely be perceived as a threat by those in positions of high power, so censorship resistance is an
essential feature. 

Unfortunately, contemporary lending/borrowering DApps on Cardano are implemented in ways that cannot
possibly create a true trustless and decentralized credit-debt market. To fully appreciate this, it
is first important to understand the deficiencies of the status quo:

### Oracles

DApps that rely on off-chain information feeds are subject to the integrity of the underlying oracle
network. It is pretty easy to find examples of faulty oracle feeds causing DeFi issues that resulted
in users losing money. Not only that, but in many cases, *oracles are simply not necessary*. As long
as users are able to negotiate the terms *before* initiating the loan, there is no need for
liquiditations, and therefore, no need for oracles. In TradFi, liquiditations are usually only used
with *margin loans* which are typically short-term loans and/or are generally only used to buy more
securities (ie, speculate on markets). In a typical economy, most loans are *not* margin loans; you
wouldn't use a margin loan to start a business.

> 
> > What if the value of the colleral falls 6 months into the loan?! Shouldn't Alice be forced to put

Reviewer: ```suggestion
An example of margin loans causing liquidation in the oracle system:

> What if the value of the colleral falls 6 months into the loan?! Shouldn't Alice be forced to put
```
(or equivalent introduction... since the quoted question doesn't refer to the previous paragraphs)

available to scripts in a transaction, is there even room for this extra logic? The composition
logic requires two things: it must support *arbitrary* compositions and it must be extremely cheap
since the logic of the actual DApps being composed must also be executed in the same transaction.
*p2p protocols do not have the same limitations since Alice personally creates and signs her own
transaction.* The above two requirements are only requirements when middle-men are creating and
executing transactions on the behalf of users. Even if some composition logic could fit in the
transaction, would it actually support arbitrary composition (as opposed to only a few priviledged
DApps being supported)? Even if the answer to this question is yes, that extra logic being executed
means less space for other DApps to compose in the transaction. Therefore, protocols that require
batchers will never be able to offer the the same level of trustless composition as fully p2p
protocols. The only way to actually match the level of composition of p2p protocols is for the
batcher based composition to *not be trustless* (without the composition logic, more DApps can fit

> in a transaction).

Reviewer: Even if not broken up into paragraphs in the original, it should be broken up here... not readable since requires too many details be accumulated in the reader's head between moving from one conceptual step to another.

They can deny sharing the credit histories for borrowers they don't like. This would effectively
nullify a borrower's entire credit history since lenders will most likely treat an unknown history
the same as no history. Even if the people with these beefy computers are honest, they are a very
centralized group that may be easy for governments target. If a government is able to commandeer
these beefy computers, governments will still be able to exert undue influence. Blockchain would
fail to be a check on government abuse. This argument applies to priorietary credit history
databases as well.

Credit histories *must* be on-chain and *must* be easy to lookup, even with an average computer.
Otherwise, the credit history database itself would be an existential risk for a DeFi credit-debt
market.

> NOTE: Data availability is still a work in progress, but so far, there are no alternatives to
> storing the critical information on-chain. There may never be an alternative; the whole point of
> the layer 1 chain is to create a global database of critical information. Custody of money is
> *not* the only critical information an economy relies on.

> ### Centralized

Reviewer: Might also add to this section this link w/ short introduction to changes already in progress: https://cointelegraph.com/news/upcoming-defi-rules-europe-could-ban-non-decentralized-protocols

Author: I added a paragraph that conncts to the link. I also added a comment that a "truly decentralized DApp should be *legally unregulatable*". I like this phrase and believe it is true even though I am not 100% sure it is true. Perhaps there is still some applicable regulation even though I can't picture what that would be...

The Loan Asset Beacon names for the negotiation phase and active phase are deliberately the same to
simplify the off-chain querying.

The borrower's credit history is directly tied to what happens with the Borrower ID Beacon.

The 2 Loan ID Beacons act as *Lock & Key NFTs*. One copy is *always* stored with the collateral
while the other is the bond NFT given to the lender. Whenever the lender wants to do something (eg,
update the payment address or claim expired collateral), the transaction must contain *both* copies.
Because of this, the Key NFT can act as a proxy for the lender's approval of the action; the Key NFT
can only move to a new output or be burned if the owner approved the transaction. Furthermore, it is
trivial for lenders to look up the current state of the loan stored with the Lock NFT since it will
have the same name as the Key NFT.

### Six Aiken Smart Contracts

Due to the amount of logic required for this protocol and the desire to minimize the impact from
redundant executions, this protocol uses 6 separate smart contracts; *not all of them are required
with each transaction*. Each smart contract is dedicated to a specific purpose. As a consequence of

> this, there are actually two separate minting policies for beacons tokens: one for negotiation phase

Reviewer: ```suggestion
this, there are actually two separate minting policies for beacon tokens: one for negotiation phase
```

  -- for the minUTxOValue.
  | UpdateLenderAddress { newAddress :: Address, depositIncrease :: Integer }
  -- | Clean up remaining beacons from Finished Active UTxOs and/or claim Lost Active UTxOs.
  | Unlock
  deriving (Generic,Show)
```

##### Loan Spending Smart Contract Datums
A DApp UTxO's datum can either be an `AskDatum`, an `OfferDatum`, or an `ActiveDatum`. `POSIXTime`
is always in milliseconds.

```haskell
-- | A type synonym for an asset's full name.
type Asset = (CurrencySymbol,TokenName)

-- | The penalty to apply whenever the minimum payment is not met in a given compounding period.
data Penalty
  = NoPenalty
  | FixedFee Integer
  | PercentFee Rational

data LoanDatum 
  = AskDatum
      -- | The policy id for the negotiation beacon script.
      { negotiationBeaconId :: CurrencySymbol
      -- | The policy id for the active beacon script.
      , activeBeaconId :: CurrencySymbol
      -- | The borrower's staking credential hash as a token name.
      , borrowerId :: TokenName
      -- | The asset to be loaned out.
      , loanAsset :: Asset
      -- | The token name for the loan asset beacon.
      , assetBeacon :: TokenName

>       -- | The size of the loan.
>       , loanPrinciple :: Integer

Reviewer: ```suggestion
      -- | The size of the loan.
      , loanPrincipal :: Integer
```
The term "Principle" (just in this machine readable term as I can see) are spelled wrong throughout the document: https://www.investopedia.com/terms/p/principal.asp

Author: This entire time I thought it was "loan principle" :upside_down_face:

Reviewer: `loanPrinciple :: String` == Why you are borrowing the money. 🤣

      , collateralIsSwappable :: Bool
      -- | The time at which the lender's claim period will expire and the Active UTxO will be
      -- considered "lost".
      , claimExpiration :: POSIXTime
      -- | The time at which the loan will expire.
      , loanExpiration :: POSIXTime
      -- | The loan's remaining unpaid balance.
      , loanOutstanding :: Rational
      -- | The total payments made this loan epoch.
      , totalEpochPayments :: Integer
      -- | The loan's unique indentifier. This is the name of the loan's Lock & Key NFTs.
      , loanId :: TokenName
      }
```

##### Payment Datum
This datum is attached to all direct outputs to lenders. It is used to prevent double-satisfaction.

```haskell
-- | The `CurrencySymbol` is always the active beacon policy id, and the `TokenName` is always
-- the Loan ID this payment output corresponds to.
newtype PaymentDatum = PaymentDatum (CurrencySymbol,TokenName)
```

The `CurrencySymbol` is included to prevent double-satisfaction during DApp composition; just using
the `TokenName` may not be enough to guarantee uniqueness.

### Ask UTxOs

> Borrower's begin negotiations by creating a new Ask UTxO with their desired terms at their loan

Reviewer: ```suggestion
Borrowers begin negotiations by creating a new Ask UTxO with their desired terms at their loan
```

`CreateCloseOrUpdateAsk`.
- The staking credential used with the `CreateCloseOrUpdateAsk` redeemer must be the one used for
the Borrower ID *and* the target loan address.
- All outputs with beacons must have exactly two kinds of beacons, with exactly one unit of each:
    - an Ask beacon with the token name "Ask"
    - a loan asset beacon with the token name corresponding to this Ask's loan asset.
- All outputs with beacons must have a valid `AskDatum`:
    - `negotiationBeaconId` == the negotiation smart contract policy id
    - `activeBeaconId` == the negotiation smart contract's hard-coded active smart contract hash
    - `borrowerId` == the staking credential hash used in the `CreateCloseOrUpdateAsk` redeemer
    - `loanAsset` == asset that corresponds to the loan asset beacon in this output
    - `assetBeacon` == this output's loan asset beacon's token name
    - `loanPrinciple` > 0
    - `loanTerm` > 0
    - `collateral` must not be empty, must be sorted lexicographically, and must *not* have any
    duplicates
- All outputs with beacons must contain at least one unit of each asset in the collateral list.

> - All outputs with beacon must not have any extraneous assets (ada is always allowed).

Reviewer: ```suggestion
- All outputs with beacons must not have any extraneous assets (ada is always allowed).
```
(parallelism with items before & after on the list)

> Both borrower credentials must still approve the transaction so there is no risk of theft. This
> works because the negotiation smart contract doesn't care where inputs with beacons come from; it
> only cares about outputs with beacons. In this case, there aren't any outputs with beacons.

##### Updating Ask UTxOs

If the Ask UTxO is changing the loan asset, the loan asset beacon will need to be changed which
means the negotiation smart contract must be executed as a minting policy. However, if no beacons
need to be changed, the negotiation smart contract can be executed as a staking script. *All other
requirements are the same as closing Ask UTxOs.* Even the redeemers used are the same.

It is *not* possible to move an Ask UTxO to a new loan address in the case where a borrower wishes
to use a different staking credential. Using both the minting execution and staking execution in a
single transaction does not help because they will both check the outputs. Whichever execution has
the old credential will complain that the new Ask UTxO output is using the wrong Borrower ID name
and is locked at the wrong loan address.

### Offer UTxOs

> Lender's continue negotiations by creating a new Offer UTxO with their desired terms at the

Reviewer: ```suggestion
Lenders continue negotiations by creating a new Offer UTxO with their desired terms at the
```

- Every offer must have a corresponding lender payment output:
    - It must be locked at the `lenderAddress` in the corresponding `ActiveDatum`
    - It must contain the `offerDeposit` from the corresponding `OfferDatum` and exactly 1 Loan ID
    beacon with the token name matching `loanId` in the corresponding `ActiveDautm`
    - A `PaymentDatum` where the `CurrencySymbol` is the active smart contract policy id and the
    `TokenName` is the `loanId` from the corresponding `ActiveDatum`
- The negotiation smart contract must burn *all* negotiation beacons attached to the Ask and Offer
UTxOs.
- The active smart contract must mint exactly:
    - 1 Active beacon for every Offer UTxO among the inputs.
    - 1 Borrower ID beacon for every Offer UTxO among the inputs.
    - 2 Loan ID beacons for every Offer UTxO among the inputs (each Offer UTxO gets their own unique
    Loan ID pair).
    - 1 loan asset beacon for every Offer UTxO accepted that requires this asset beacon.
- The staking credential of the loan address where all inputs come from must approve the
transaction.

*Valid Offers and Asks are those containing the proper beacons.*

> The loan's start time is set to the invalid-before bound of the acceptance transaction's validity
> interval. If any offers have an expiration set, then the invalid-hereafter bound will also be needed
> to prove the offer has not expired. If no offers have expirations set, the invalid-hereafter bound
> can be dropped.

Reviewer: ```suggestion
The loan's start time is set to the `invalid-before` bound of the acceptance transaction's validity
interval. If any offers have an expiration set, then the `invalid-hereafter` bound will also be needed
to prove the offer has not expired. If no offers have expirations set, the `invalid-hereafter` bound
can be dropped.
```
... to clarify that these are fixed parameters of the transaction? (used in other places as well).  Maybe not needed if audience is only the Cardano community.

- All Active UTxOs being unlocked must come from the same loan address.
- If any of the Active UTxOs spent with `Unlock` contain beacons, the active smart contract must be
executed as a minting policy using the `BurnRemainderOrUnlockLost` redeemer.
- For all Active UTxOs being unlocked:
    - If the Borrower ID is still present, the loan's `claimExpiration` must have passed (using
    invalid-before).
    - All active beacons among the inputs must be burned.
- The loan address' staking credential must signal approval.

The invalid-before bound is used to prove the claim expiration has passed. When unlocking multiple
loans, invalid-before can be set to the latest claim expiration. If all Active UTxOs in the
transaction are Finished Active UTxOs, the invalid-before bound is not necessary.

Any lost collateral UTxOs claimed with this redeemer will still count as defaults for the borrower.

Invalid Active UTxOs (ie, UTxOs with an `ActiveDatum` but no beacons) can be spent using just the
`Unlock` spending redeemer. The loan spending script will directly check for the borrower's
approval.

## Benchmarks and Fee Estimations (YMMV)

> **No CIPs or hard-forks are needed. This protocol works on the Cardano blockchain, as is.**

Reviewer: I'm really happy to hear this & it answered a question I had while reading about some CIPs I'd reviewed on the following subjects, and so was reassured this works with current Cardano architecture:
- input / output ordering in transactions (as mentioned is important for payments / interest / update outputs)
- transaction composability
- any other changes to Plutus and/or the Ledger that might be needed for a full, efficient implementation

cost for this. Every part of the loan is tracked, not just whether it was a default. For example,
lenders can see how much of the loan a borrower paid back before they defaulted. Therefore, not all
defaults (should) impact creditworthiness in the same way. Futhermore, when new versions of the
protocol come out, the borrower's credit history is not lost; a borrower's credit history transcends
the protocol's version (assuming they are using the same credential across versions).

Consider what this means for the non-banked people who live in countries unable to support their own
credit history databases due to lack of money, corruption, etc. By using Cardano-Loans, the
non-banked can trustlessly start building their own credit histories for as little as 0.3 ADA per loan
action. This likely results in a credit history database that is 1,000-10,000x cheaper to run than
the credit history systems in developed nations like the United States. And despite being
astronomically cheaper to maintain, it offers things not even the contemporary databases offer:

- Identity theft is *cryptographically impossible*.
- Credit histories are globally accessible and verifiable.

> Using Cardano-Loans, the non-banked can finally leap-frog passed the incompetence/corruption of
> their own nations and build up their own, (possibly) globally recognized, economic identity.

Reviewer: ```suggestion
Using Cardano-Loans, the non-banked can finally leap-frog passed the incompetence/corruption of
their own national financial systems and build up their own, (possibly) globally recognized, economic identity.
```
Permits the possibility that a nation might not be corrupt but still have a corrupt (manipulative, coercive, repressive) financial system.

### Negotiable Collateralization

Since the lenders explicitly set the relative values of each collateral asset used, there is no need
to use oracles or enable liquidations (these are not meant to be margin loans). However, there is a
risk for the value of the collateral to fall precipitously during the duration of the loan. If the
collateral ever falls too low in value, **this is the lender's fault**. The lender either should
have asked for a higher relative rate, or a different asset all together. It is up to the lender to
use proper risk-management.  If people want to take back control from governments, they need to
accept more responsibility. *Freedom isn't free.*

> ### Under-Collateralized Loans

Reviewer: Somewhere in this section, can you add something... a small sub-heading, maybe... to explain how the elegant & persuasive explanations here would still hold true if volatility swings were to suddenly make a lot of properly collateralised loans undercollateralised... including any situations long-term lender "reputation" might be wrecked involuntarily by huge volatility swings?

Or, if this problem can't be avoided, explain why it is normal & should be considered an acceptable risk?  In either case I think this high-level overview would suit non-technical people (company execs, managers) considering this system vs. some more centralised alternatives.

Reviewer 2: I am coming around to your way of thinking about not needing to continually price the collateral.  Instead shifting the burden to the lenders to properly price the risk.

reputation. The brand with the better reputation may have 1000x more revenue than the other brand,
despite selling the same exact item for the same exact price.

People factor in reputation whenever they decide who to interact with. It takes a very long time to
build up a good enough reputation for people to want to do business with you.

> It takes 20 years to build a reputation and five minutes to ruin it. If you think about that,
> you'll do things differently.  
> -- Warren Buffett

A lot of people forget that *reputation is an asset*; it is just an intangible one. When a borrower
takes out a loan, their reputation is also applied as collateral to the loan. So would Bob walk away
from a loan where he only needed to back 80% of it with tangible assets? The answer depends on how
much his reputation is worth to him. If his reputation is worth less than the remaining 20%, Bob
will likely walk away. However, if his reputation is worth more than the remaining 20%, Bob will
likely pay back the loan to protect the value of his reputation. And just like with all assets, the
value of a reputation depends on its scarcity. **Whether the reputation is for a public,

> pseudonymous, or anonymous identity is irrelevent** - the dark web marketplaces are proof of this.

Reviewer: ```suggestion
pseudonymous, or anonymous identity is irrelevant** - the dark web marketplaces are proof of this.
```

they would first look up all credentials linked to the target one. Then, they would query all of the
credentials' credit histories. Nothing would need to change for Cardano-Loans.

### Version Compatibility for Negotiations

Currently, a borrower and a lender need to use the same protocol version to negotiate and enforce a
loan. There are some technical difficulties with enabling cross-version compatibility:

- If they use different datums, which payment credential would be used?
- The different versions may not know about each others' beacons. How would proper minting/burning
of the beacons be enforced?

These questions are currently unanswered which is why this feature still requires R&D.

## Conclusion

The Cardano-Loans protocol is a rethinking of how Cardano's DeFi economy could evolve. It forgoes
reliance on global/external token prices in favor of incentivizing the creation of a CSL-native p2p
credit-debt market. It is censorship-resistant, scalable, and straightforward in its design. Wallets
and other front-ends can easily integrate the protocol into their UI, and even provide all necessary

> API query services. Endogenous price and rate discovery empowers users to create their own economy,
> one that is decoupled from the existing financial system, in pursuit of something more fair, equal,
> and accessible to all.

Reviewer: ```suggestion
API query services. Endogenous price and rate discovery empowers users to create their own economy —
one that is decoupled from the existing financial system — in pursuit of something more fair, equal,
and accessible to all.
```
(these are em dashes) Former version may also be grammatically correct, but considering this is the last sentence & most extraordinary conclusion then I think it helps for it to be punctuated like a speaker was reading it.

# Cardano-Loans

> A p2p-DeFi lending/borrowering protocol with an on-chain credit history. Lenders and borrowers can

Reviewer 2: borrowing

trustlessly negotiate all loan terms prior to initiating the loan. Once the loan starts, the
protocol will enforce all terms agreed upon during negotiations.

> As with all [p2p-DeFi protocol](https://github.com/zhekson1/CSL-DeFi-Protocols), all users maintain

Reviewer 2: protocols

    - [Term Extensions and Re-Negotiations](#term-extensions-and-re-negotiations)
    - [Linkable Credit Histories](#linkable-credit-histories)
    - [Version Compatibility for Negotiations](#version-compatibility-for-negotiations)
- [Conclusion](#conclusion)
    
## Abstract

Cardano-Loans is a *fully p2p* lending/borrowing protocol for the Cardano Settlement Layer (CSL). It
empowers users to create and operate a CSL-native credit-debt market via trustlessly negotiable and
enforcable p2p loans, and on-chain credit history. This circumvents the need for oracles and/or
concentrated lending pools in favor of *endogenous* price and interest-rate discovery. Users deploy
(and interact with each others') script addresses, so full spending *and* delegation control of
assets are maintained, and contract upgrades occur democratically. This protocol natively supports
*all* Cardano assets as both loan assets and collateral assets - even assets that have yet to be
created.

## Motivation

A healthy credit-debt market is a vital (and often the largest) component of a thriving economy. A
healthy market is one in which prices reflect underlying reality as quickly and faithfully (with as

> little friction) as possible. For the sake of avoiding a whole treatise on economics, suffice to say

Reviewer 2: This is really a personal style issue and therefore you may not agree, but I think it might read better with commas instead of (brackets) in a few areas.  IE: Users deploy and interact with each others' script addresses.  ... as quickly, and faithfully, with as little friction as possible.

Author: I agree with the first two but the last one is meant to be a definition of "faithfully". Therefore, for that one, I'll change that one to: "(ie, with as little friction)".

credit-debt market that seeks to be an alternative to the current global financial system will
likely be perceived as a threat by those in positions of high power, so censorship resistance is an
essential feature. 

Unfortunately, contemporary lending/borrowering DApps on Cardano are implemented in ways that cannot
possibly create a true trustless and decentralized credit-debt market. To fully appreciate this, it
is first important to understand the deficiencies of the status quo:

### Oracles

DApps that rely on off-chain information feeds are subject to the integrity of the underlying oracle
network. It is pretty easy to find examples of faulty oracle feeds causing DeFi issues that resulted
in users losing money. Not only that, but in many cases, *oracles are simply not necessary*. As long
as users are able to negotiate the terms *before* initiating the loan, there is no need for
liquidations, and therefore, no need for oracles. In TradFi, liquidations are usually only used
with *margin loans* which are typically short-term loans and/or are generally only used to buy more
securities (ie, speculate on markets). In a typical economy, most loans are *not* margin loans; you

> wouldn't use a margin loan to start a business.

Reviewer 2: I like your framing for why "cardano-loans" here and I don't want you to change the document because this is obviously the philosophy of cardano-loans.

However, I don't agree.  I think that all loans are heading in the direction of being treated as margin loans, and this is particularly so since the 2008 GFC.  For example, even the current value of US treasuries held by banks, guaranteed to redeem at full value, had to be manipulated by the FED with their recent BTFP program.  IE: Even those "guaranteed by Govt" assets caused bank liabilities to act like margin loans.  I think we are gradually heading into a new world where every asset is valued constantly, even houses when they become tokenised and fractionally saleable.

Author: >  I think we are gradually heading into a new world where every asset is valued constantly, even houses when they become tokenised and fractionally saleable.

I agree that society is trending this way, but I am not convinced it will be successful. I think there are physical limits to how continuous real-time price discovery can be, especially on blockchain. For example, network bandwidth probably cannot support all of the traffic required to support real-time price discovery of everything; every time someone updates a price, the network needs to transmit that difference to everyone. Day-traders can accidentally ddos the network. 

Then, you have the blockchain bandwidth of constantly updating the state on-chain using some kind of oracle. Assuming there is actually a good solution to oracles, can they update the prices fast enough to keep up while still allowing network space for other activities? I personally wouldn't bet on this.

A lot of people believe oracles are required for DeFi because of the following:

> What if the value of the colleral falls 6 months into the loan?! Shouldn't Alice be forced to put
> up more collateral?

No, she shouldn't. Bob, her lender, knew the loan was going to be a year long loan. If he was
concerned about the value of the collateral falling during the life of the loan, he could have asked
for more collateral and/or asked for different collateral. The fact that cryptocurrency markets can
be extremely volatile does not matter. Volatile assets also exist in TradFi (eg, small cap stocks,
emerging market assets, etc), and this is exactly how TradFi manages the risk. If the collateral

> ever falls too low, relative to the outstanding loan, **it is the lender's fault**. The borrower

Reviewer 2: TradFi uses margin loans when stocks/shares are the collateral.

Author: I agree that is *usually* the case. I do not personally work on wall street so I don't have concrete evidence, but I believe this is not always true. I think it depends on the volatility of stocks/shares. A small cap stock will likely either be rejected as collateral or they will want to use margin loans. Meanwhile a large cap stock with low volatility may be accepted as collateral without margin loans (also accounting for the loan's duration). I don't think this approach is open to the average person, but millionaires and billionaires likely do get deals like this. 

I can change the sentence to: "and this is one of the methods TradFi uses to manage the risk."

This disconnect is guaranteed to cause market inefficiencies...

Furthermore, in order for a new asset to be made available as either the loan asset or a collateral
asset, new liquidity pools must be created with that asset. For some DApps, this requires a
"governance" action while others just require enough users to deposit the required assets into the
new pool. Either way, this is bad. Imagine if a new nation formed and created a new currency on
the blockchain. Either this new nation needs permission for its currency to join the DeFi ecosystem,
or it needs to be popular enough to attract enough capital to start a new liquidity pool. The former
gives incredible power to the gate keepers of new assets while the latter creates a chicken-or-egg
problem: the new currency will only be popular if it is useful for DeFi but it won't be useful for
DeFi unless it is actually listed from being popular. It shouldn't matter what any third-parties
think about an asset. If Alice wants to borrower the new nation's currency and Bob is willing to
lend it to her, why can't they use this currency? It shouldn't matter how many other people are
willing to use those assets.

> Finally, **high TVL is a competition to see who can be most inefficient with DeFi capital.**

Reviewer 2: Yes!!! Love it!!!

Capital-at-rest is wasted economic potential. A high TVL fundamentally means the supply is more than
the demand. The higher the TVL, the greater the disconnect between the supply and demand. This is a
huge misallocation of resources. If the DeFi economy is going to reach its full potential, **TVL

> should be as low as possible.**

Reviewer 2: 100%

immediately execute it; this all takes place in a single transaction.

How would you do the above composition in TradFi? Just the first step of converting DJED to ADA
would likely require a few days for the trade to settle. In the meantime, you are exposed to the
volatility of ADA. Then, if your broker doesn't support the secondary market, you need to transfer
your ADA to another broker which would take another few days to settle. Buying the options contract
may take ANOTHER few days to settle. Do you see where this is going? By the time all of the settling
has finished, the opportunity may have long passed.

While the above example doesn't include Cardano-Loans, the point is that composition allows for
creating truly complex economic transactions from very simple building blocks. Just like with
Wolfram's cellular automaton, it is very hard to predict what the ultimate economy will look like.
Because of this, **all core DeFi DApps should priortize composability**. People may come up with
useful compositions that the DApp developers never considered. The more compositions that are
possible, the larger, and more complex the DeFi economy can become. And as the risk-management

> example shows, composability can make the complex DeFi economy more resilient.

Reviewer 2: 100%

### No Trustless Credit History

You simply cannot have a healthy credit-debt market without credit history. However, if a market
does have credit history but that history is not trustless, it just creates another attack vector
governments and hackers can use to target the growing DeFi economy. 

Some DApps try to rely on off-chain credit history databases, but this is not a good approach.
*Reading* credit histories must be just as trustless as *writing* credit histories. Even if the
blockchain records the credit history trustlessly, if lenders can't query the credit history, it is
effectively the same as there being no credit history. A centralized database, or a database that is
hard to read, is easy to commandeer/censor. 

In practical terms, if only beefy computers are able to read the credit histories from the
blockchain, those with these computers are in a priviledged position between borrowers and lenders.

> They can deny sharing the credit histories for borrowers they don't like. This would effectively

Reviewer 2: They might even do the same for borrowers they DO like in order to stifle competition from other lenders.

nullify a borrower's entire credit history since lenders will most likely treat an unknown history
the same as no history. Even if the people with these beefy computers are honest, they are a very
centralized group that may be easy for governments target. If a government is able to commandeer
these beefy computers, governments will still be able to exert undue influence. Blockchain would

> fail to be a check on government abuse. This argument applies to priorietary credit history

Reviewer 2: proprietary

databases as well.

Credit histories *must* be on-chain and *must* be easy to lookup, even with an average computer.
Otherwise, the credit history database itself would be an existential risk for a DeFi credit-debt
market.

> NOTE: Data availability is still a work in progress, but so far, there are no alternatives to
> storing the critical information on-chain. There may never be an alternative; the whole point of
> the layer 1 chain is to create a global database of critical information. Custody of money is
> *not* the only critical information an economy relies on.

### Centralized

Contemporary lending/borrowing DApps use liquidity pools which means users must sacrifice custody,
delegation control, and voting control of their assets in order to participate in DeFi. As these

> DApps gain adaption, the underlying Cardano blockchain becomes less decentralized and secure.

Reviewer 2: adoption

Furthermore, despite issuing "governance tokens", most of these DApps are controlled by nothing more
than a multisig. The DApp developers have the ability to freeze all market activity. The claim is
that these DApps will eventually be decentralized, but it remains to be seen if this is even
possible with their designs. Most layer 1 blockchains are still struggling to achieve
decentralization (it is a very hard problem). In the meantime, it is trivial for governments and
hackers to target these DApps since they only really need to target the multisig key holders.

Malicious actors are not the only risk to DeFi: regulations are also an existential risk since they
are another way for governments to exert influence on the DeFi economy. The regulatory landscape is
still changing, but they all center around whether the DApp is actually decentralized ([Europe
example](https://cointelegraph.com/news/upcoming-defi-rules-europe-could-ban-non-decentralized-protocols)).
If there is even a single centralization bottleneck (eg, closed-source smart contracts, permissioned
batchers, minting of yield tokens, multisig control, etc), the entire DApp can be regulated. If a

> DApp is truly decentralized, it should be *legally unregulatable*.

Reviewer 2: I like the way you worded that.

Not all blockchain applications need to be decentralized, but core DeFi functionality *definitely

> needs to be decentralized.* 

Reviewer 2: 100%

- **A Single DApp Address for each Borrower** - every borrower gets their own personal DApp address.
All negotiations involving the borrower take place in this address and all active loans live in this
address. This makes it very easy for front-ends to integrate Cardano-Loans as well as for borrowers
to manage the stake/voting power for assets in the DApp.
- **Efficient Off-chain Sorting of Loans** - by using beacon tokens, users can easily query the
protocol for any information they may want. These beacon queries can even be combined to create more
complicated queries. For example, lenders can see all current Asks and optionally filter them by
loan asset and/or offered collateral.
- **Democratic Upgradability** - users decide if and when to upgrade to a new version of the DApp.
No one can decide this for them, or force them to move to a new version (eg, by freezing all
functionality for the current version).
- **Full Composability** - this protocol is natively composable with all other DeFi DApps. It can
even be composed with itself: a borrower can accept an offer in the same transaction where they make

> a payment.

Reviewer 2: Just a general comment about trade-offs with restriction-less financial instrument features:

Whilst in keeping with a permissionless defi system, this design choice impacts composability in the sense that it becomes more difficult to collect similar financial instruments into baskets.

But maybe you are right.  Maybe things will go this way because AI may make pricing such varied instruments easier and thus a large parcel composed of such instruments could be more easily valued???  TradFi markets re-package loans into these huge collateralied debt obligations so that they can trade fractionally for greater liquidity, and so that large institutional lenders can participate within their regulatory bounds.

Maybe that could be another project for you?  Developing software to accurately price bundles of these cardano-loans and re-package them into CDOs which are in-turn divided up into different tranches to meet the requirements of large lenders?

Author: The CDOs was one of the applications I had in mind with this design. I am not personally interested in creating software for them, but I do think there is value in someone doing it and I think there is enough financial incentive for it to eventually happen.

> Whilst in keeping with a permissionless defi system, this design choice impacts composability in the sense that it becomes more difficult to collect similar financial instruments into baskets.

I don't think this is true since the beacon tokens make filtering the entire DApp market very easy. For the CDO example, it is theoretically easy to cross-reference borrowers' credit histories with the target active loans; it would just require some moderately powerful hardware.

## Specification

This section is the low-level specification for the protocol. If you are only interested in the
high-level aspects, feel free to skip to the next [section](#benchmarks-and-fee-estimations-ymmv).

### The Borrower's Address

Borrowers each create a unique "loan" address - this is where all negotiations take place *and*
where all collateral is kept until the loan is either repaid or expired. As is common in
*distributed dApps*, all such loan addresses use the same validator script for the payment
credential, and a unique, user-defined staking key or script for the staking credential.
Owner-related actions are delegated to the staking credential *by* the validator script, so the user

> maintains full control of all assets at the address. The borrower's credit history is

Reviewer 2: Full control?  Presumably they can't spend their tokens.

Author: Full *custody* is probably better. They can't spend locked collateral, but they are the only entity that possibly could.

cryptographically tied to the staking credential used in the loan address.

Since negotiations occur in the borrower's address, the borrower maintains staking and voting rights
over all assets throughout the life cycle of the loan. This includes collateral, which makes sense,
since collateral belongs to the borrower until they default. It also includes any assets attached to

> offers, which incentivizes lenders to be proactive and not leave offers open for too long. 

Reviewer 2: Does this mean it is possible for a potential borrower to create an address, deposit some Ada to it and continue to receive staking rewards (so no ongoing cost).  Then simply wait for loan offers but not respond, and benefit from the increased staking yield when potential lenders send stake-able tokens as a loan offer?

Author: Yes, but I don't see this as an issue for two reason:
1. Ask UTxOs are timestamped. If this attack is actually an issue, lenders can just avoid old Ask UTxOs.
2. The Cardano epoch is 5 days. If a borrower is *actually* looking for a loan, why would it take 5 days to accept a lender's offer? Besides, if lenders want to protect their staking rewards, they can easily close all of their offers right before the end of the epoch and open them again after the next snapshot. In this scenario, the borrower wouldn't benefit at all from this attack. Front-ends can also remind lenders to do this.

of the payments ([calculator](https://www.nerdwallet.com/calculator/loan-calculator)). The
additional 700 ADA is due to the interest. This 1700 ADA is Mike's investment for an 8 year credit
history of 1000 ADA.

Mike needs to ask Alice for a loan of about 1000 ADA. It would be a red-flag if he suddenly asked
for 10,000 ADA despite his history only dealing with loans 10x smaller than that. Even in TradFi,
credit limits are slowly increased as a borrower's credit history matures. Let's be generous and say
Alice agrees to a 1000 ADA loan where only 80% of the loan's value is backed by tangible assets.
Therefore, if Mike walked away with the 1000 ADA, he would net 200 ADA. This 200 ADA is his return.
In otherwords, his 1700 ADA investment yielded a total of 200 ADA over an 8 year time period. That
equates to an average yearly return of 1.4% (`200 profit / 1700 investment / 8 years * 100`). Mike
put in all that time and effort farming the identities for only a 1.4% APY. But remember, Mike did
*not* get 1.4% APY each year for 8 years, he had to wait the full 8 years to see any return from

> this attack.

Reviewer 2: Except it doesn't actually cost Mike anything because he is making the payments to himself through his multiple fake identities.  The cost to him is time and transaction fees.

Still I am coming around to agree that under collateralised loans will be made because lenders can check addresses of previous lenders and compare with other lenders they have verified.  This way they could be more certain that the previous loans were real.

Author: > Except it doesn't actually cost Mike anything because he is making the payments to himself through his multiple fake identities. The cost to him is time and transaction fees.

I agree there is barely any cost, but that is why I was trying to frame it as invested capital vs return. A loan for 1000 ADA still requires the 1000 ADA. However, perhaps the velocity of the 1000 ADA could be high enough to keep the invested capital fixed :thinking:. If it is possible to use the same 1000 ADA to maintain loans for 10,000 ADA loans, this would make this attack much more feasible.

I may remove this section entirely. I still believe it is possible to have a healthy under-collateralized loan market, but I now don't think my current thinking is correct. Cross-referencing other lenders like you said, or linking borrower IDs to public identities (even without DIDs) may be enough.

Author: I removed the under-collateralized loan discussion section and elaborated on the previous [section](https://github.com/fallen-icarus/cardano-loans/tree/v1.0.0.0rc?tab=readme-ov-file#negotiable-collateralization). I just focused on the data analyses one could use to identify fake identities, such as check past lenders.

Cardano-Loans is designed to be independent from the traditional financial system, in favor of
endogenously producing its own. As such, every piece of a loan, including the relative values of
assets to collateral, interest rates, and term length are all negotiated and agreed upon fully p2p -
no oracle feeds required. 

Although this presents a bootstrapping problem, it may be overcome by the fact that this protocol
may be the fastest way for the most financially underserved peoples to begin building a credit
history. Prospective borrowers are incentivized to build p2p relationships within a global
marketplace of prospective lenders, who themselves are incentivized to lend by the relatively high
rates that financially underserved borrowers would be willing to pay. Since offers are public,
lenders can see all offers being made to the borrower and/or for that loan asset, resulting in a
healthy competition between lenders, and a healthy interest-rate discovery mechanism. And since this
protocol does not rely on any team or "utility token", this market is immune to any top-down
influence.

With enough users and liquidity, this protocol may eventually *serve* as the de-facto oracle for

> market-driven interest-rate discovery.

Reviewer 2: Love it.  Lofty goals, but possible.  I would definitely love to see that eventuate.

version by then.

### DID Compatibility

The current design may already be compatible with DIDs. If a government either allowed a user to
register their Cardano staking credential or issued them a dedicated one, this protocol already
natively supports DIDs. Lenders would just need to cross-reference the government's registered
credentials when they query a borrower's credit history. However, if governments used NFTs instead
of credentials as DIDs, the protocol may not already support DIDs.

Therefore, whether any changes are needed for Cardano-Loans to support DIDs entirely depends on how
DIDs are issued.

### Term Extensions and Re-Negotiations

Refinancing loans are a useful economic feature, and sometimes, borrowers just need a little more
time to pay back a loan. These features were not implemented because they still require R&D. It is
unclear whether these features need to be supported directly by Cardano-Loans or whether, thanks to
its composability, a separate DeFi protocol will suffice. Using a separate protocol, that "plugs-in"
to Cardano-Loans may result in less overall complexity than putting all of the features into one

> protocol.

Reviewer 2: I don't think you need anything extra.  Why can't this just be a negotiation between the lender and borrower to pay off the previous loan and simultaneously open a new loan?  This would essentially roll the loan.  Or even a negotiation with a new lender to pay off the existing loan and generate a new one.  Cardano-loans can already do that.

Author: You currently cannot accept an offer in the same transaction where another loan is fully paid off because the beacon mint/burn count will not agree (partial payments are fine since there is no required mint/burn). The ideal way to make them agree is to have a single count, but this requires a single script to tally up the required beacon mint/burn count for all actions. Because of all of the logic required, it doesn't fit into a single script. Breaking it into multiple scripts introduces security vulnerabilities as well as some redundant executions. I have not found a method I am content with.

There are also questions that need to be addressed:
- How does a borrower signal their desire to refinance/re-negotiate? Dedicated beacons will likely be required.
- How does the lender know the borrower will use the new loan to pay off the old one? There is currently no guarantee for this.

Reviewer 2: I don't understand the internal smart contract limitations well enough, but two things come to mind:
1. If there is a limit on beacon burning/creation caused by completely closing the loan, then maybe the lender could mandate paying off 99.99% such that it is then much more advantageous for the borrower to subsequently completely close out this last tiny bit to get the loan "off the books" and improve his credit score.
2. I was under the impression that when a lender signs a transaction that this can be composed of a chain of transactions such that they all atomically succeed or fail, enabled by inherent UTxO determinism.  Why can't the lender sign a transaction which amounts to a new loan as well as paying off (or 99.99%) the old loan?

Author: Paying off 99.99% could enable the composition, but currently, the borrower is the only person who can spend the loan UTxO in order to make a payment on it so it would ultimately have to be the borrower who submits the final refinancing transaction. It is also the borrower who accepts offers. The lender is not part of either of these transactions; they just create the Offer UTxOs.

With the way the contracts currently work, if a lender created a new offer with the intention of it being used to pay off 99.99% of a previous loan, how does the lender know the new principal will actually be used like this? Right now, there would be nothing stopping the borrower from accepting the second loan while still keeping the first. While this may increase the chances of the borrower defaulting, I don't think this is enough of an incentive for the borrower to actually use the new principal to pay off 99.99% of the old loan. New logic would be needed to enforce the proper usage of this new principal.

If lenders are fine with the risks, the protocol does currently support the borrower accepting the new offer and paying off 99.99% of the old loan in the same transaction. But there is currently no way for either the borrower or lender to broadcast their intention for refinancing. If the borrower created a new Ask UTxO for the same amount, it would just look like the borrower wants another loan. How would lenders tell the difference between a borrower who wants a second loan for 1000 ADA and another borrower who wants to refinance a loan for 1000 ADA? There is currently no way for lenders to distinguish between the two types of Asks. The same is true for Offers. Dedicated beacons would need to be added in order to distinguish between the two types of Asks/Offers.
